File: Binary Search/splitArray.js
const splitArray = (nums, k) => {
  const check = (minLargestSum) => {
    let currentSum = 0;
    let subArrayCount = 0;

    for (let num of nums) {
      if (currentSum + num > minLargestSum) {
        subArrayCount++;
        if (subArrayCount > k) return false;
        currentSum = 0;
      }
      currentSum += num;
    }

    if (subArrayCount + 1 <= k) return true;

    return false;
  };

  let left = -Infinity,
    right = 0;
  for (let num of nums) {
    left = Math.max(num, left);
    right += num;
  }

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (check(mid)) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return left;
};

const nums = [1, 2, 3, 4, 5],
  k = 1;

console.log(splitArray(nums, k));

File: Binary Search/SplitArray.java
public class SplitArray {
  public int splitArray(int[] nums, int k) {
    int left = Integer.MIN_VALUE;
    int right = 0;
    for (int num : nums) {
      left = Math.max(num, left);
      right += num;
    }

    int minLargestSum = 0;
    while (left <= right) {
      int mid = left + (right - left) / 2;
      if (check(mid, nums, k)) {
        right = mid - 1;
        minLargestSum = mid;
      } else {
        left = mid + 1;
      }
    }

    return minLargestSum;
  }

  private boolean check(int maxSum, int[] nums, int k) {
    int currentSum = 0;
    int subArrayCount = 0;
    for (int num : nums) {
      if (currentSum + num > maxSum) {
        currentSum = 0;
        subArrayCount++;
        if (subArrayCount > k)
          return false;
      }
      currentSum += num;
    }

    return subArrayCount + 1 <= k;
  }
}
